### PR TITLE
add a demo of content filter listener

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
Signed-off-by: Chen Lihui <lihui.chen@sony.com>

### CODE REVIEW

**Comment by iuhilnehc-ynos** [comment]
@asorbini @MiguelCompany 

As we know, the content filter has operations as below,
`RelOp   ::=    ‘=’ | ‘>’ | ‘>=’ | ‘<’ | ‘<=’ | ‘<>’ | like``

I expected that the keyword `like` can work as https://www.w3schools.com/sql/sql_ref_like.asp.
Unfortunately, it seems [it](https://github.com/ros2/demos/blob/183f7cd9c0e4979bb33e6586c941cfc9e2d1f14b/demo_nodes_cpp/src/topics/listener_content_filter.cpp#L48) doesn't work.

Is there a limitation in the DDS?

**Comment by MiguelCompany** [comment]
@iuhilnehc-ynos On Fast DDS it should work as expected. See the test cases [here](https://github.com/eProsima/Fast-DDS/blob/master/test/unittest/dds/topic/DDSSQLFilter/DDSSQLFilterTests.cpp#L1305)

RTI Connext, according to [this comment](https://github.com/ros2/design/pull/282#issuecomment-1080942322), does not support `like`

**Review by ivanpauno — APPROVED** [review]
LGTM!

Could you add a test [here](https://github.com/ros2/demos/tree/master/demo_nodes_cpp/test)?

**Review by fujitatomoya — APPROVED** [review]

**Comment by fujitatomoya** [comment]
Once this demo is merged, i will link this demo to Humble Release Note. (see https://github.com/ros2/ros2_documentation/pull/2396)

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16555/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11142/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16934/)

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (wjwwood) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: Maybe we could use the `MATCH` feature? Something like `"data MATCH %0"` and `{"'Hello World: \d*[02468]'"}` might match only even numbers?

**Comment by ivanpauno** [comment]
> @iuhilnehc-ynos On Fast DDS it should work as expected. See the test cases [here](https://github.com/eProsima/Fast-DDS/blob/master/test/unittest/dds/topic/DDSSQLFilter/DDSSQLFilterTests.cpp#L1305)
> 
> RTI Connext, according to https://github.com/ros2/design/pull/282#issuecomment-1080942322, does not support like

Could we modify the example to work with both?

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (iuhilnehc-ynos) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: `match` is not standard in the DDS spec.

What about using `data = 'Hello World: 1' or data = 'Hello World: 10'`? What I thought is that it seems not a good experience for the user, so I insist on using the `like`.

Comment (wjwwood) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: Match isn't supported? That seems pretty unfortunate. 

This brings up another point. We need to define what can and cannot be used in the filter expressions at the ros level. I don't think it's a good thing to just reference what the underlying middleware supports only.

And I don't think we should use an example that only works with one vendor.

Comment (wjwwood) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: So while the "like" example is more compelling, the fact that it is not portable kind of disqualifies it in my mind.

Comment (iuhilnehc-ynos) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: To add the limitation at the ros level, I think it's not fair to FastDDS.

I'll try to use `match` instead of `like` at this PR.

Comment (fujitatomoya) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: > match is not standard in the DDS spec.

did you check this `match` does not work? it seems Fast-DDS supports it based on the source code DDSSQLFilter.

Comment (iuhilnehc-ynos) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: > did you check this match does not work? 

No, not yet. I was afraid it's not supported in other coming DDS that might not support `match` based on the DDS spec.

Comment (iuhilnehc-ynos) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: connextdds(fnmatch) and fastdds(std::regex) use different regular expression. `fnmatch` doesn't know the `{"'Hello World: \\d*[02468]'"}`, which works well in fastdds(std::regex).

Comment (iuhilnehc-ynos) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: I'll update content filter just with `"data = %0", {"'Hello World: 10'"}`.

I believe that there will some new issues to ask `why can't I use 'like' for content filter setting?`.

Comment (wjwwood) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:48]: To be honest we shouldn't encourage the use of ones outside the standard then (unless we support them at the rcl layer as a fallback) because we cannot use "like" in any meaningful way if it's not required to be supported by the rmw implementation. 

I think your suggestion of matching one exact string is fine. Really the chatter example just isn't well suited to this kind of content filtering it seems.

**Comment by iuhilnehc-ynos** [comment]
@ivanpauno 

> Could you add a test [here](https://github.com/ros2/demos/tree/master/demo_nodes_cpp/test)?

Sorry, I don't know how to add a test for this listener_content_filter, it seems the current `test_executables_tutorial.py.in` can't work as I expect.

I expect that running with `colcon test --packages-select demo_nodes_cpp --ctest-args -R test_tutorial_talker_listener_content_filter__rmw_cyclonedds_cpp` should fail with [the patch](https://github.com/ros2/demos/files/8478295/0001-test.patch.txt), but it doesn't.

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (ivanpauno) [inline demo_nodes_cpp/src/topics/listener_content_filter.cpp:52]: Could we add a log here saying "subscribed to topic X with filter expression Y"?
So users understand a bit more when running the example.

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (ivanpauno) [inline demo_nodes_cpp/src/topics/listener_content_filter.cpp:49]: Maybe the number (currently `10`) could be configurable by the user.
Maybe that makes the demo more interesting (?).
Feel free to ignore the comment though.

**Comment by ivanpauno** [comment]
> I expect that running with colcon test --packages-select demo_nodes_cpp --ctest-args -R test_tutorial_talker_listener_content_filter__rmw_cyclonedds_cpp should fail with [the patch](https://github.com/ros2/demos/files/8478295/0001-test.patch.txt), but it doesn't.

Yes, that's not going to work because the test is waiting for the output in the file to be printed https://github.com/ros2/demos/blob/7a15d0bf14e5765ea223873b6927be908c7acbb3/demo_nodes_cpp/test/test_executables_tutorial.py.in#L55, and that happens as well if no content filter is applied.
You'll need to adapt the test, or adapt the file so it doesn't match the output if "Content filter is not enabled since it's not supported" was logged before (I'm not sure if that's possible).

---

Please don't share a patch, they are hard to understand without being applied to the code to have more context.
It's better to push the commit and ask the question, as github gives more context (or if you don't want to push in this branch, you can do that in a new branch/fork and share the link to the github diff).

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (fujitatomoya) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp:49]: I actually agree with this comment. probably string chatter example is interesting enough to show this feature. how about filtering specific range of data? that would be one of the most common case?

**Comment by fujitatomoya** [comment]
@iuhilnehc-ynos all comments are addressed, right? can you confirm? so that we can ask final review to @ivanpauno @wjwwood

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (wjwwood) [inline demo_nodes_cpp/src/topics/listener_content_filter.cpp:49]: I'm not 100% sure how `.arguments()` works off the top of my head, but couldn't contain more options potentially? It looks fragile to assume that if the size is 2 then it's the arguments you expect...

There's also no usage docs or anything.

Maybe you could just use a ROS parameter instead? Or just not make this configurable from the cli?

**Review by wjwwood — COMMENTED** [review]
Just curious why not put this kind of thing in `ros2/examples` instead?

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (iuhilnehc-ynos) [reply demo_nodes_cpp/src/topics/listener_content_filter.cpp]: > Just curious why not put this kind of thing in `ros2/examples` instead?

If put it in `ros2/examples`, I need to add an extra `talker_content_filter.cpp`, which seems unnecessary.

> Maybe you could just use a ROS parameter instead? Or just not make this configurable from the cli?

Considering it's located in `demo_nodes_cpp/src/topics`, and the concept of `ROS parameter` is only used in `demo_nodes_cpp/src/parameters`, I'll take second suggestion to remove the `options.arguments`.

**Review by ivanpauno — APPROVED** [review]
LGTM with @wjwwood approval

**Comment by fujitatomoya** [comment]
@wjwwood friendly ping.

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (wjwwood) [inline demo_nodes_cpp/src/topics/listener_content_filter.cpp:61]: This is really fragile, since if someone updated the expression and forgot to update this is would be wrong. Violates DRY.

Can we print out the filter expression and parameters from the `content_filter_options` directly instead?

File: demo_nodes_cpp/src/topics/listener_content_filter.cpp
Comment (wjwwood) [inline demo_nodes_cpp/src/topics/listener_content_filter.cpp:43]: Is this comment block needed?

**Comment by wjwwood** [comment]
w.r.t. putting it in examples. I think having something here in demos is fine so long as we have a corresponding page about it or the executable is used in some documentation or tutorial. For example, the intra-process demos have this page which shows how to run it, what you should expect to see when you run it, and **how it demonstrates** that the intra-process feature is working:

https://github.com/ros2/ros2_documentation/blob/rolling/source/Tutorials/Intra-Process-Communication.rst

If we had something similar then this executable makes sense to me. Otherwise, I'd still say examples makes more sense. And I don't think you need a talker equivalent if it is put in examples, the purpose there is to show how to use the API, and it's not part of a tutorial or demo document that tells you to run it and observe some effect.

I hope that distinction makes sense.

**Comment by fujitatomoya** [comment]
@wjwwood @ivanpauno 
CC: @iuhilnehc-ynos 

i just created https://github.com/ros2/examples/pull/341.

how about the following, sounds like a plan?

- we could keep this demo. but this demo could be adjusted more `demonstration` perspective. IMO, i would use float or integer to simulate the sensing date, and filtering the specific range. this would be common use case and easy to get a picture how it works. any comments are welcome.
- https://github.com/ros2/ros2_documentation/pull/2396 can link this demonstration.
- adding tutorial `how to use`, and that doc links to https://github.com/ros2/examples/pull/341

**Comment by wjwwood** [comment]
> https://github.com/ros2/ros2_documentation/pull/2396 can link this demonstration.
> adding tutorial how to use, and that doc links to https://github.com/ros2/examples/pull/341

I'd flip this around so that:

* https://github.com/ros2/ros2_documentation/pull/2396 links to the example
* a new `how to use` "tutorial" and have that doc link to this demo application (and use `talker` from `demo_nodes_cpp` as part of the "how to use")
* After it is in place, https://github.com/ros2/ros2_documentation/pull/2396 could also link to the tutorial/demo doc

We could set up a new example that uses floats or something like you said, but that would be for the demo/tutorial combo and not the minimal example.

**Comment by fujitatomoya** [comment]
@iuhilnehc-ynos 

> We could set up a new example that uses floats or something like you said, but that would be for the demo/tutorial combo and not the minimal example.

can we address this? current code looks good to us, just adjusting data type for using float to simulate the situation to filtering the sensing data. for doing that, we need to add publisher as well.

how about following?

- emergency temperature demonstration
  - say publisher is publishing temperature data with [std_msgs/msg/Float64](https://github.com/ros2/common_interfaces/blob/master/std_msgs/msg/Float64.msg). make a loop to publish data -100 to +150 F in order with certain offset.
  - subscription use content filtering (if it is supported), and only receives data with `“data < %0 OR data > %1”` with `{"-30.0", "100.0"}`, then print emergency temperature.

i believe demonstrating with expected common use case would be much easier to get the picture of this feature for user.

**Comment by iuhilnehc-ynos** [comment]
@ros-pull-request-builder retest this please

File: demo_nodes_cpp/CMakeLists.txt
Comment (fujitatomoya) [inline demo_nodes_cpp/CMakeLists.txt:75]:
```suggestion
  src/topics/content_filtering_publisher.cpp
  src/topics/content_filtering_subscriber.cpp
```

sorry i was not clear on this, i think files, classes and components can be general names for content filtering, so that user can tell these are related to content filtering.

**Review by fujitatomoya — APPROVED** [review]
implementation looks good to me.

**Review by ivanpauno — APPROVED** [review]
Example looks really nice!
Thanks for the work!

**Comment by wjwwood** [comment]
Do you guys have demo documentation to go with this? Something along the lines of this document: https://docs.ros.org/en/rolling/Tutorials/Intra-Process-Communication.html

**Comment by fujitatomoya** [comment]
@wjwwood i will be back with PR for that, let me have a couple of days.

**Comment by wjwwood** [comment]
Sure, no problem, I would just hold this until that is ready.

**Comment by fujitatomoya** [comment]
@wjwwood @ivanpauno @iuhilnehc-ynos 
CC: @clalancette 

i just opened https://github.com/ros2/ros2_documentation/pull/2441, could you guys take a look at to give me some feedback?

**Comment by fujitatomoya** [comment]
either @wjwwood or @clalancette 

can you merge this with https://github.com/ros2/ros2_documentation/pull/2441 at the same time?

**Comment by fujitatomoya** [comment]
@ivanpauno thanks! all green, can you merge this?

**Comment by fujitatomoya** [comment]
@clalancette can you merge this? i already did https://github.com/ros2/ros2_documentation/pull/2441

**Comment by fujitatomoya** [comment]
this is only required to backport humble package.

**Comment by wjwwood** [comment]
@mergifyio backport humble

**Pull request merged**
